apps/emqx/src/emqx_limiter/src/emqx_limiter_shared/emqx_limiter_shared.erl

```diff
@@ -61,101 +98,264 @@
+                limiter_id => LimiterId,
+                amount => Amount,
+                result => Result
+            }),
+            {Result, State1}
     end.
 
 -spec put_back(client_state(), non_neg_integer()) -> client_state().
-put_back(LimiterId = State, Amount) ->
-    case emqx_limiter_bucket_registry:find_bucket(LimiterId) of
-        undefined ->
-            State;
-        #{counter := Counter, index := Index} ->
-            ok = put_back(Counter, Index, Amount),
-            State
-    end.
+put_back(#{bucket_ref := BucketRef, mode := Mode} = State, Amount) ->
+    case Mode of
+        #token_mode{us_per_token = UsPerToken} ->
+            #{last_time_aref := LastTimeARef, last_time_index := LastTimeIndex} = BucketRef,
+            ok = atomics:sub(LastTimeARef, LastTimeIndex, UsPerToken * Amount);
+        #mini_token_mode{} ->
+            #{mini_tokens_cref := MiniTokensCRef, mini_tokens_index := MiniTokensIndex} = BucketRef,
+            AmountMini = Amount * 1000,
+            ok = counters:add(MiniTokensCRef, MiniTokensIndex, AmountMini)
+    end,
+    State.
+
+-spec inspect(client_state()) -> map().
+inspect(#{bucket_ref := BucketRef, mode := Mode, options := Options} = _State) ->
+    #{last_time_aref := LastTimeARef, last_time_index := LastTimeIndex} = BucketRef,
+    LastTimeUs = atomics:get(LastTimeARef, LastTimeIndex),
+    #{mini_tokens_cref := MiniTokensCRef, mini_tokens_index := MiniTokensIndex} = BucketRef,
+    MiniTokens = counters:get(MiniTokensCRef, MiniTokensIndex),
+    #{capacity := Capacity, interval := IntervalMs} = Options,
+    TimeLeftUs = now_us_monotonic() - LastTimeUs,
+    #{
+        mode => Mode,
+        time_left_us => TimeLeftUs,
+        mini_tokens => MiniTokens,
+        tokens_left => ((TimeLeftUs * Capacity) div IntervalMs) div 1000
+    }.
 
 %%--------------------------------------------------------------------
 %%  Internal functions
 %%--------------------------------------------------------------------
 
-start_child(Sup, ChildSpec) ->
-    case supervisor:start_child(Sup, ChildSpec) of
-        {ok, _} ->
-            ok;
-        {error, _} = Error ->
-            Error
+make_buckets(Names, MiniTokensCRef, LastTimeARef, NowUs) ->
+    make_buckets(Names, MiniTokensCRef, LastTimeARef, NowUs, 1, []).
+
+make_buckets([], _MiniTokensCRef, _LastTimeARef, _NowUs, _Index, Res) ->
+    lists:reverse(Res);
+make_buckets(
+    [{Name, Options} | Names], MiniTokensCRef, LastTimeARef, NowUs, Index, Res
+) ->
+    BucketRef = #{
+        mini_tokens_cref => MiniTokensCRef,
+        mini_tokens_index => Index,
+        last_time_aref => LastTimeARef,
+        last_time_index => Index,
+        last_burst_time_index => Index + 1
+    },
+    Mode = calc_mode(Options),
+    ok = apply_burst(Mode, Options, BucketRef, NowUs),
+    make_buckets(Names, MiniTokensCRef, LastTimeARef, NowUs, Index + 2, [
+        {Name, BucketRef} | Res
+    ]).
+
+try_consume(State0, Options, Amount, NowUs) ->
```

Comment:
maybe use a different name, e.g. `try_consume_base` to differentiate from `try_consume/2`